File: src/state/player.ts
import { signal } from '@preact/signals';
import { Block } from '../tzx/types';
import { renderTape, playbackTimeline, playbackOrder, positionAt, TSTATES_PER_SEC } from '../tzx/audio';
import { Side, audioMode, setStatus } from './store';

export const playing = signal(false);
/** Which pane's tape is playing, so only its list shows the marker. */
export const playingSide = signal<Side | null>(null);
/** Block index (in the tape) currently being played, or -1. */
export const playingBlock = signal(-1);
/** Elapsed and total seconds of the current playback; both 0 when idle. */
export const playPos = signal({ elapsed: 0, total: 0 });

let ctx: AudioContext | null = null;
let source: AudioBufferSourceNode | null = null;
let frame = 0;

export function playBlocks(blocks: Block[], indices?: number[], side: Side | null = null) {
  stopPlayback();
  if (!ctx) ctx = new AudioContext();
  if (ctx.state === 'suspended') void ctx.resume(); // browsers start it suspended until a user gesture
  const rate = ctx.sampleRate;
  const order = indices ?? playbackOrder(blocks);
  const samples = renderTape(blocks, { sampleRate: rate, mode: audioMode.value }, order);
  if (samples.length === 0) {
    setStatus('Nothing to play');
    return;
  }
  const buffer = ctx.createBuffer(1, samples.length, rate);
  buffer.copyToChannel(samples as Float32Array<ArrayBuffer>, 0);
  const node = ctx.createBufferSource();
  source = node;
  node.buffer = buffer;
  node.connect(ctx.destination);
  node.onended = () => {
    if (source === node) stopPlayback();
  };
  const startAt = ctx.currentTime;
  node.start(startAt);
  const total = samples.length / rate;
  const { starts } = playbackTimeline(blocks, order);
  playing.value = true;
  playingSide.value = side;
  playingBlock.value = order[0] ?? -1;
  playPos.value = { elapsed: 0, total };
  setStatus('');
  const tick = () => {
    if (source !== node || !ctx) return;
    // Output latency is not exposed by every WebKit; fall back to the context clock alone.
    const latency = (ctx as AudioContext & { outputLatency?: number }).outputLatency ?? 0;
    const elapsed = Math.max(0, Math.min(total, ctx.currentTime - startAt - latency));
    const at = order[positionAt(starts, elapsed * TSTATES_PER_SEC)] ?? -1;
    if (at !== playingBlock.value) playingBlock.value = at;
    playPos.value = { elapsed, total };
    frame = requestAnimationFrame(tick);
  };
  frame = requestAnimationFrame(tick);
}

export function stopPlayback() {
  cancelAnimationFrame(frame);
  if (source) {
    const node = source;
    source = null;
    try { node.stop(); } catch { /* already stopped */ }
  }
  playing.value = false;
  playingSide.value = null;
  playingBlock.value = -1;
  playPos.value = { elapsed: 0, total: 0 };
}

File: desktop/src/player.rs
//! Playback, the port of `src/state/player.ts`.
//!
//! Web Audio takes a whole `AudioBuffer` and plays it; cpal takes a callback and
//! asks for frames. The rendered tape therefore lives in an `Arc<[f32]>` that
//! the audio thread reads and a cursor the UI thread watches — the same three
//! numbers the web build exposes as signals (`playing`, `playingBlock`,
//! `playPos`), read here as one `Progress`.
//!
//! The tape is rendered at the device's own sample rate, so nothing resamples.

use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::Arc;

use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
use cpal::{Error as CpalError, SampleFormat};
use tapeti_core::audio::{RenderOptions, TSTATES_PER_SEC};
use tapeti_core::types::Block;

use crate::state::Side;
use crate::tape::{playback_timeline, position_at, render_tape};

/// What the status bar and the block list read while a tape plays.
#[derive(Clone, Copy, Default)]
pub struct Progress {
    pub playing: bool,
    /// Which pane's tape is playing, so only its list shows the marker.
    pub side: Option<Side>,
    /// Block index in the tape, or -1.
    pub block: i32,
    pub elapsed: f64,
    pub total: f64,
}

struct Shared {
    /// Samples handed to the device so far.
    at: AtomicUsize,
    done: AtomicBool,
}

pub struct Player {
    /// Dropping the stream stops playback; kept only while something plays.
    stream: Option<cpal::Stream>,
    shared: Arc<Shared>,
    side: Option<Side>,
    order: Vec<u32>,
    starts: Vec<u64>,
    rate: u32,
    total_samples: usize,
    block: i32,
}

impl Default for Player {
    fn default() -> Self {
        Player {
            stream: None,
            shared: Arc::new(Shared { at: AtomicUsize::new(0), done: AtomicBool::new(true) }),
            side: None,
            order: Vec::new(),
            starts: Vec::new(),
            rate: 44_100,
            total_samples: 0,
            block: -1,
        }
    }
}

impl Player {
    pub fn playing(&self) -> bool {
        self.stream.is_some() && !self.shared.done.load(Ordering::Relaxed)
    }

    /// Called once a frame: advances the block marker and notices the end of the
    /// tape, the way the `requestAnimationFrame` tick does on the web.
    pub fn poll(&mut self) -> Progress {
        if self.stream.is_none() {
            return Progress::default();
        }
        if self.shared.done.load(Ordering::Relaxed) {
            self.stop();
            return Progress::default();
        }
        let at = self.shared.at.load(Ordering::Relaxed).min(self.total_samples);
        let elapsed = at as f64 / f64::from(self.rate);
        let tstates = (elapsed * f64::from(TSTATES_PER_SEC)) as u64;
        self.block = self.order.get(position_at(&self.starts, tstates)).map_or(-1, |i| *i as i32);
        Progress {
            playing: true,
            side: self.side,
            block: self.block,
            elapsed,
            total: self.total_samples as f64 / f64::from(self.rate),
        }
    }

    /// Render `order` of `blocks` and start playing it. Returns what to put in
    /// the status bar when there is nothing to play or no device to play on.
    pub fn play(
        &mut self,
        blocks: &[Block],
        order: Vec<u32>,
        side: Option<Side>,
        mic: bool,
    ) -> Option<String> {
        self.stop();
        let host = cpal::default_host();
        let Some(device) = host.default_output_device() else {
            return Some("No audio output device".into());
        };
        let config = match device.default_output_config() {
            Ok(c) => c,
            Err(e) => return Some(format!("No audio output configuration: {e}")),
        };
        let rate = config.sample_rate();
        let channels = config.channels() as usize;
        let opts = RenderOptions { sample_rate: rate, mic, amplitude: 1.0 };
        let samples: Arc<[f32]> = render_tape(blocks, opts, &order).into();
        if samples.is_empty() {
            return Some("Nothing to play".into());
        }
        let timeline = playback_timeline(blocks, &order);

        let shared = Arc::new(Shared { at: AtomicUsize::new(0), done: AtomicBool::new(false) });
        let stream = match build_stream(&device, &config, samples.clone(), channels, shared.clone()) {
            Ok(s) => s,
            Err(e) => return Some(format!("Cannot play: {e}")),
        };
        if let Err(e) = stream.play() {
            return Some(format!("Cannot play: {e}"));
        }
        self.total_samples = samples.len();
        self.rate = rate;
        self.order = order;
        self.starts = timeline.starts;
        self.side = side;
        self.shared = shared;
        self.stream = Some(stream);
        self.block = -1;
        None
    }

    pub fn stop(&mut self) {
        self.stream = None;
        self.side = None;
        self.block = -1;
        self.order.clear();
        self.starts.clear();
        self.total_samples = 0;
    }
}

fn build_stream(
    device: &cpal::Device,
    config: &cpal::SupportedStreamConfig,
    samples: Arc<[f32]>,
    channels: usize,
    shared: Arc<Shared>,
) -> Result<cpal::Stream, CpalError> {
    let err = |e| eprintln!("audio: {e}");
    let cfg: cpal::StreamConfig = config.config();
    // One mono buffer fanned out to every channel; `next` is the frame cursor
    // the UI reads back as elapsed time.
    macro_rules! stream {
        ($t:ty, $conv:expr) => {{
            let shared = shared.clone();
            let samples = samples.clone();
            let conv: fn(f32) -> $t = $conv;
            device.build_output_stream(
                cfg.clone(),
                move |out: &mut [$t], _: &cpal::OutputCallbackInfo| {
                    let mut at = shared.at.load(Ordering::Relaxed);
                    for frame in out.chunks_mut(channels) {
                        let v = samples.get(at).copied().unwrap_or(0.0);
                        for s in frame.iter_mut() {
                            *s = conv(v);
                        }
                        at += 1;
                    }
                    shared.at.store(at, Ordering::Relaxed);
                    if at >= samples.len() {
                        shared.done.store(true, Ordering::Relaxed);
                    }
                },
                err,
                None,
            )
        }};
    }
    match config.sample_format() {
        SampleFormat::F32 => stream!(f32, |v| v),
        SampleFormat::I16 => stream!(i16, |v| (v.clamp(-1.0, 1.0) * 32767.0) as i16),
        SampleFormat::U16 => stream!(u16, |v| ((v.clamp(-1.0, 1.0) * 32767.0) as i32 + 32768) as u16),
        SampleFormat::I32 => stream!(i32, |v| (f64::from(v.clamp(-1.0, 1.0)) * 2_147_483_647.0) as i32),
        other => Err(CpalError::with_message(
            cpal::ErrorKind::InvalidInput,
            format!("unsupported sample format {other:?}"),
        )),
    }
}
